lib/Conversion/TorchToLinalg/TorchToLinalg.cpp

```diff
@@ -1251,29 +1251,61 @@ class ConvertAtenLinearOp : public OpConversionPattern<AtenLinearOp> {
 };
 } // namespace
 
-
-static Value promoteScalarToDtype(OpBuilder &b, Location loc, Value scalar, Type dtype) {
-  // TODO: For the integer case, we probably need the unconverted dtype to
+// Convert a scalar value to the target type. The scalar value can be an element
+// from a tensor or a scalar in the pytorch dialect. Both the scalar and dtype
+// should be converted builtin types.
+static Value convertScalarToDtype(OpBuilder &b, Location loc, Value scalar,
+                                  Type dtype) {
+  Type scalarType = scalar.getType();
+  if (scalarType == dtype)
+    return scalar;
+
+  // TODO: For the byte(ui8) or char(i8) case, we need the unconverted dtype to
   // be able to know if we need signed or unsigned conversion.
-  if (dtype.isa<mlir::FloatType>()) {
-    if (scalar.getType().isa<mlir::FloatType>()) {
-      // `scalar` will always be f64 since that is what the TypeConverter
-      // converts !torch.float to.
-      return b.create<arith::TruncFOp>(loc, scalar, dtype);
-    } else {
-      assert(scalar.getType().isa<mlir::IntegerType>());
-      // `scalar` will always be i64 since that is what the TypeConverter
-      // converts !torch.int to.
-      return b.create<arith::SIToFPOp>(loc, scalar, dtype);
+  auto isByteOrChar = [](Type type) {
+    if (auto integerTy = type.dyn_cast<mlir::IntegerType>()) {
+      return integerTy.getWidth() == 8;
     }
+    return false;
+  };
+
+  if (isByteOrChar(scalarType) || isByteOrChar(dtype) ||
+      scalarType.isSignlessInteger(1) || dtype.isSignlessInteger(1)) {
+    // TODO: Handle bool type.
+    mlir::emitError(loc)
+        << "unsupported byte, char or bool type for convertScalarToDtype "
+        << scalarType << "(scalar type) -> " << dtype << "(dtype)";
+    return nullptr;
   }
-  mlir::emitError(loc) << "promoteScalarToDtype for dtype " << dtype;
-  return nullptr;
+
+  if (auto dtypeFloat = dtype.dyn_cast<mlir::FloatType>()) {
+    if (auto scalarFloat = scalarType.dyn_cast<mlir::FloatType>()) {
+      if (scalarFloat.getWidth() > dtypeFloat.getWidth())
+        return b.create<arith::TruncFOp>(loc, scalar, dtype);
+      // Only scalarFloat width < dtypeFloat width can reach here.
+      return b.create<arith::ExtFOp>(loc, scalar, dtype);
+    }
+    assert(scalarType.isa<mlir::IntegerType>());
+    return b.create<arith::SIToFPOp>(loc, scalar, dtype);
```

Comment:
add comment: this is safe because ui8/si8 are the only ones where unsigned handling is needed, and we checked for that case above.